### PR TITLE
Support other integer types for SubstringUTF8 & RightUTF8 functions

### PR DESCRIPTION
**Pull request by gengliqi**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/9473

Problem Summary:

### What is changed and how it works?

```commit-message
Support other integer types for SubstringUTF8 & RightUTF8 functions
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsString.cpp [hunk 1]
```diff
@@ -1683,13 +1683,21 @@ class FunctionSubstringUTF8 : public IFunction
                   using StartType = std::decay_t<decltype(start_type)>;
                   // Int64 / UInt64
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:1684]: this comment is out of date now

File: dbms/src/Functions/tests/gtest_substring.cpp [hunk 2]
```diff
@@ -27,9 +27,160 @@ class SubString : public DB::tests::FunctionTest
...
+            sub_string.executeFunction(
+                "substringUTF8",
+                createColumn<Nullable<String>>({"www.pingcap.com", "ww.pingcap.com", "w.pingcap.com", ".pingcap.com"}),
+                createConstColumn<T1>(4, 1),
+                createColumn<T2>({1, 2, 3, 4})));
+    }
+};
+
+template <typename T1, typename T2>
+class TestConstLength
```
Comment (yibin87) [inline dbms/src/Functions/tests/gtest_substring.cpp:149]: Please make sure we cover const nullable cases.

Comment (gengliqi) [reply dbms/src/Functions/tests/gtest_substring.cpp:149]: `TestAllIntTypes` and `TestAllUIntTypes` consist of nullable types.

File: dbms/src/Functions/FunctionsString.cpp [hunk 3]
```diff
@@ -1841,48 +1866,67 @@ class FunctionSubstringUTF8 : public IFunction
...
-            return {true, start_field.get<UInt64>()};
+            static_assert(
+                std::is_same_v<Integer, UInt8> || std::is_same_v<Integer, UInt16> || std::is_same_v<Integer, UInt32>
+                || std::is_same_v<Integer, UInt64>);
+            return {true, val};
         }
     }
 
     template <typename F>
     static bool getNumberType(DataTypePtr type, F && f)
```
[the remark this answers is not in the document]
Comment (gengliqi) [reply dbms/src/Functions/FunctionsString.cpp:1919]: I changed it to `FunctionSubstringUTF8::getValueFromLengthColumn`

**Review by windtalker — APPROVED** [review]
lgtm

**Pull request merged**